hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/StorageBasedLockProvider.java

```diff
@@ -0,0 +1,581 @@
+      // 3. Corner case 2: lock is expired but no one has acquired it yet, lock
+      // renewal "revived" the expired lock.
+      // Please note we expect the corner cases almost never happens.
+      // Action taken for corner case 2 is just a best effort mitigation. At least it
+      // prevents further data corruption by
+      // letting someone else acquire the lock.
+      Pair<LockUpdateResult, StorageLockFile> currentLock = this.lockService.tryCreateOrUpdateLockFileWithRetry(
+          () -> new StorageLockData(false, System.currentTimeMillis() + lockValidity, ownerId),
+          getLock(),
+          LOCK_UPSERT_RETRY_COUNT);
+      switch (currentLock.getLeft()) {
+        case ACQUIRED_BY_OTHERS:
+          logger.error("Owner {}: Unable to renew lock as it is acquired by others.", ownerId);
+          // No need to extend lock lease anymore.
+          return false;
+        case UNKNOWN_ERROR:
+          // This could be transient, but unclear, we will let the heartbeat continue
+          // normally.
+          // If the next heartbeat run identifies our lock has expired we will error out.
+          logger.warn("Owner {}: Unable to renew lock due to unknown error, could be transient.", ownerId);
+          // Let heartbeat retry later.
+          return true;
+        case SUCCESS:
+          // Only positive outcome
+          this.setLock(currentLock.getRight());
+          logger.info("Owner {}: Lock renewal successful. The renewal completes {} ms before expiration for lock {}.",
+              ownerId, oldExpiration - System.currentTimeMillis(), lockFilePath);
+          // Let heartbeat continue to renew lock lease again later.
+          return true;
+        default:
+          throw new HoodieLockException("Unexpected lock update result: " + currentLock.getLeft());
+      }
+    } catch (Exception e) {
+      logger.error("Owner {}: Exception occurred while renewing lock", ownerId, e);
+      return false;
+    }
+  }
+
+  // ---------
+  // Utilities
+  // ---------
+
+  /**
+   * Method to calculate whether a timestamp from a distributed source has
+   * definitively occurred yet.
+   */
+  protected boolean isCurrentTimeCertainlyOlderThanDistributedTime(long epoch) {
+    return System.currentTimeMillis() > epoch + CLOCK_DRIFT_BUFFER_MS;
+  }
+
+  private String buildLockObjectPath(String lockFolderName, String lockTableFileName) {
+    // Normalize inputs by removing trailing slashes
+    // We know lockTableFileName has already been parsed.
+    if (lockFolderName.startsWith("/")) {
+      lockFolderName = lockFolderName.substring(1);
+    }
+
+    // Append a slash only if one isn't already present.
+    return lockFolderName + (lockFolderName.endsWith("/") ? "" : "/") + lockTableFileName + ".json";
+  }
+
+  private String slugifyLockFolderFromBasePath(String basePathKey) {
+    // Remove the prefix once
+    String cleanedPath = basePathKey.replaceFirst("^(gs://|s3://|s3a://)", "");
```

Comment:
should we get the list of supported storage schemes from StorageSchema eums only?

Reply:
removing this entirely